Docs: Dev FAQ - About indirect conditions (#3698)
ArchipelagoMW/Archipelago: docs/apworld_dev_faq.md

```py
item_pool += [self.create_filler() for _ in range(total_locations - len(item_pool))]
```

---

### I learned about indirect conditions in the world API document, but I want to know more. What are they and why are they necessary?

The world API document mentions indirect conditions and **when** you should use them, but not *how* they work and *why* they are necessary. This is because the explanation is quite complicated.

> Region sweep (the algorithm that determines which regions are reachable) is a Breadth First Search of the region graph from the Menu region, checking entrances one by one and adding newly reached nodes (regions) and their entrances to the queue until there is nothing more to check.

Reviewer: ```suggestion
Region sweep (the algorithm that determines which regions are reachable) is a Breadth-First Search of the region graph from the Menu region, checking entrances one by one and adding newly reached nodes (regions) and their entrances to the queue until there is nothing more to check.
```

> Region sweep (the algorithm that determines which regions are reachable) is a Breadth First Search of the region graph from the Menu region, checking entrances one by one and adding newly reached nodes (regions) and their entrances to the queue until there is nothing more to check.

Reviewer: Not necessarily "Menu" anymore

> For performance reasons AP only checks every entrance once. However, if entrance access conditions depend on regions, then it is possible for this to happen:

Reviewer: ```suggestion
For performance reasons, AP only checks every entrance once. However, if entrance access conditions depend on regions, then it is possible for this to happen:
```

Reviewer 2: is that comma proper grammar? imo it reads better without it

Reviewer: It is, although looking into a few different guidebooks, it seems there's disagreement on whether short prepositional phrases ("short" being two, three, four, or five words, depending on the guidebook) actually need a comma. I see what you mean though, feel free to ignore this comma

> For performance reasons AP only checks every entrance once. However, if entrance access conditions depend on regions, then it is possible for this to happen:

Reviewer 2: ```suggestion
For performance reasons AP only checks every entrance once. However, if any entrance access condition depends on regions, then it is possible for this to happen:
```

1. An entrance that depends on a region is checked and determined to be nontraversable because the region hasn't been reached yet during the graph search.
2. After that, the region is reached by the graph search. The entrance *would* now be determined to be traversable if it were rechecked.

> To account for this case, AP would have to recheck all entrances every time a new region is reached, until no new regions are reached.

Reviewer 2: ```suggestion
To account for this case, AP would have to recheck all entrances every time a new region is reached until no new regions are reached.
```

Author: Is this comma actually correct?

Reviewer: Removing the comma is fine, it's also technically fine to keep. The sentence is kind of awkward but I'm not really thinking of anything better

However, there is a way to **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it. This is what an indirect condition is.

> This keeps almost all of the performance upsides. Even a game making heavy use of indirect conditions (See: The Witness) is still way way faster than if it just blanket "rechecked all entrances until nothing new is found".

Author: There is "For performance reasons, ..." some paragraphs before this
Might not be direct enough, idk

```py
item_pool += [self.create_filler() for _ in range(total_locations - len(item_pool))]
```

---

### I learned about indirect conditions in the world API document, but I want to know more. What are they and why are they necessary?

The world API document mentions indirect conditions and **when** you should use them, but not *how* they work and *why* they are necessary. This is because the explanation is quite complicated.

Region sweep (the algorithm that determines which regions are reachable) is a Breadth-First Search of the region graph from the origin region, checking entrances one by one and adding newly reached nodes (regions) and their entrances to the queue until there is nothing more to check.

> For performance reasons, AP only checks every entrance once. However, if entrance access conditions depend on regions, then it is possible for this to happen:

Reviewer 2: rewriting the suggestion because it technically had a conflict lol
```suggestion
For performance reasons, AP only checks every entrance once. However, if an entrance's access condition depends on regions, then it is possible for this to happen:
```
```suggestion
For performance reasons, AP only checks every entrance once. However, if any entrance access condition depends on regions, then it is possible for this to happen:
```
```suggestion
For performance reasons, AP only checks every entrance once. However, if entrance Access Conditions depend on regions, then it is possible for this to happen:
```
```suggestion
For performance reasons, AP only checks every entrance once. However, if entrance access_conditions depend on regions, then it is possible for this to happen:
```
giving a couple more options too, because I don't think there's one way to fix this, but as-is it reads strange without the assumption that "access condition" is one idea and not multiple words you have to parse their role in the sentence
easiest is probably to rewrite the whole sentence, but i think any of these help 😅

> 1. An entrance that depends on a region is checked and determined to be untraversable because the region hasn't been reached yet during the graph search.

Reviewer 3: ```suggestion
1. An entrance that depends on a region is checked and determined to be nontraversable because the region hasn't been reached yet during the graph search.
```
English sucks lol
Non- means something along the lines of negating, null, etc. Nontraversable means it cannot be traversed.
Un- means basically the opposite of it. Untraversable means it cannot be traversed, but with a more permanent feel? Like, it's the opposite of traversable -- it will make you go the other way (in a vibes sense).
look it's dumb but I think the change makes sense and it makes me happy

> 2. After that, the region is reached by the graph search. The entrance *would* now be determined as traversible if it were rechecked.

Reviewer: ```suggestion
2. After that, the region is reached by the graph search. The entrance *would* now be determined to be traversable if it were rechecked.
```

To account for this case, we would have to recheck all entrances every time a new region is reached, until no new regions are reached.

Because most games do not check for region access inside of entrance access conditions, AP has decided to **eschew this rechecking** and just checks every entrance once. This gives a significant performance gain to AP as a whole, about 30%-50%.

However, because some games *did* start using things like `region.can_reach` inside entrance access conditions, we provided a way to **manually** define that a *specific* entrance needs to be rechecked during region sweep if a *specific* region is reached during it. This is what an indirect condition is.
This keeps almost all of the performance upsides. Even a game making heavy use of indirect conditions (See: The Witness) is still way way faster than if it just blanket "rechecked all entrances until nothing new is found".

> The reason `location.can_reach` and `entrance.can_reach` are also affected is simple: They call `region.can_reach` on their respective parent/source region.

Reviewer: I think this part should be expanded on a bit, since it's unclear what exactly this implies/refers to
```suggestion
The reason entrance access rules using `location.can_reach` and `entrance.can_reach` are also affected is simple: They call `region.can_reach` on their respective parent/source region.
```

> We recognize it's a pretty bad beginner's trap (heck, not even a "beginner's" trap, just a trap - even for experienced AP devs), and some games are very complex with their access rules.

Reviewer 3: ```suggestion
We recognize it can feel like a trap since it will not alert you when you are missing an indirect condition, and that some games have very complex access rules.
```
More succinct, imo

> There is an open Pull Request that makes this behavior optional via a world class attribute: [Core: Region handling customization](https://github.com/ArchipelagoMW/Archipelago/pull/3682)

Reviewer: ```suggestion
There is an open Pull Request that makes this behavior optional via a world class attribute: [Core: Region handling customization](https://github.com/ArchipelagoMW/Archipelago/pull/3682).
```
